Fix/Update System.Diagnostics.Tests.ProcessTests/ProcessStart_UseShellExecuteTrue_OpenMissingFile_Throw

maryamariyan opened the pull request:
Relates to #27547

I added this PR because I saw a failure in PR https://github.com/dotnet/corefx/pull/27415:
https://mc.dot.net/#/user/maryamariyan/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/103ec78a5875598ab5c876a9a7520615796dcbf4/workItem/System.Diagnostics.Process.Tests/analysis/xunit/System.Diagnostics.Tests.ProcessTests~2FProcessStart_UseShellExecuteTrue_OpenMissingFile_Throws

The test ```ProcessStart_UseShellExecuteTrue_OpenMissingFile_Throws``` had a Process.Start statement which wasn't wrapped with using statement so the Process is not disposed right away.

stephentoub commented on src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -943,8 +943,8 @@ public void StartInfo_WebPage()
-
-            Process.Start(info); // Returns null after navigating browser
+            // Returns null after navigating browser
+            using (var p = Process.Start(info)) { }
```
I'm confused.  The comment says that this returns null.  Why do we then need a using?  Shouldn't we instead Assert.Null?

stephentoub commented on src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1087,8 +1087,8 @@ public void ShellExecute_Nano_Fails_Start()
-
-            Assert.Throws<PlatformNotSupportedException>(() => Process.Start(info));
+            
+            Assert.Throws<PlatformNotSupportedException>(() => { using (var p = Process.Start(info)) { } });
```
Using isn't needed here; you'll never get to the Dispose because Start throws. (And if it doesn't throw, then the test will fail.

stephentoub commented on src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1119,7 +1119,8 @@ public void StartInfo_BadVerb(bool useShellExecute)
             };
 
-            Assert.Equal(ERROR_FILE_NOT_FOUND, Assert.Throws<Win32Exception>(() => Process.Start(info)).NativeErrorCode);
+            Win32Exception e = Assert.Throws<Win32Exception>(() => { using (var p = Process.Start(info)) { } });
```
Same

stephentoub commented on src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1141,7 +1142,8 @@ public void StartInfo_BadExe(bool useShellExecute)
                 expected = ERROR_SUCCESS;
 
-            Assert.Equal(expected, Assert.Throws<Win32Exception>(() => Process.Start(info)).NativeErrorCode);
+            Win32Exception e = Assert.Throws<Win32Exception>(() => { using (var p = Process.Start(info)) { } });
```
Same... will stop commenting on these.

stephentoub commented on src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:
```diff
@@ -68,7 +68,7 @@ public void ProcessStart_UseShellExecute_OnLinux_ThrowsIfNoProgramInstalled()
             {
                 Console.WriteLine($"None of the following programs were installed on this machine: {string.Join(",", s_allowedProgramsToRun)}.");
-                Assert.Throws<Win32Exception>(() => Process.Start(new ProcessStartInfo { UseShellExecute = true, FileName = Environment.CurrentDirectory }));
+                Assert.Throws<Win32Exception>(() => { using (var p = Process.Start(new ProcessStartInfo { UseShellExecute = true, FileName = Environment.CurrentDirectory })) { } });
```
Also, if for some reason we really do want to call Dispose in the case where the test fails, we should instead do the Dispose outside of the Throws rather than inside it; otherwise, it could actually be the Dispose that's throwing the exception and causing a false negative on the test passing.

stephentoub commented on src/System.Diagnostics.Process/tests/ProcessTests.cs:
```diff
@@ -1636,15 +1637,16 @@ public void Process_StartWithArgumentsTest()
+                Assert.Same(password, p.StartInfo.Password);
+                Assert.Equal(domain, p.StartInfo.Domain);
+                p.Kill();
+            }
```
These last two look good.

danmosemsft commented:
@maryamariyan do you know what caused ProcessStart_UseShellExecuteTrue_OpenMissingFile_Throws  to start to fail?

maryamariyan commented:
@dotnet-bot test Outerloop Alpine.3.6 x64 Debug Build
@dotnet-bot test Outerloop CentOS.6 x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Windows x86 Debug Build
@dotnet-bot test Outerloop NETFX x64 Debug Build

maryamariyan commented:
@danmosemsft I'm not sure about the cause yet. 

In my other PR (#27415) in result of the last two times we triggered outerloop, I saw this failure but not on the first attempt.
Seems like this test does not fail all the time, at least not for me locally at the moment.

danmosemsft commented:
I see no hits from the finalizer in these results.

danmosemsft commented:
If you agree, please merge.

maryamariyan commented:
@danmosemsft as discussed offline I merged PR https://github.com/dotnet/corefx/pull/27415.

maryamariyan commented:
@dotnet-bot test Outerloop Alpine.3.6 x64 Debug Build
@dotnet-bot test Outerloop CentOS.6 x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Windows x86 Debug Build
@dotnet-bot test Outerloop NETFX x64 Debug Build

maryamariyan commented:
As per @stephentoub 's comment I removed the using statements wrapping process start that I had added initially within the Assert Throws statements.

I tested locally on ubuntu and windows and the test didn't fail. but from what I remember the test failure on the CI happened for centos, opensuse and red hat. 
I'm going to see if I can reproduce on red hat next meanwhile the outerloop is running on the CI again.

maryamariyan commented:
This outerloop failure was masked by #27426

maryamariyan commented:
@dotnet-bot test Outerloop Alpine.3.6 x64 Debug Build
@dotnet-bot test Outerloop CentOS.6 x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Windows x86 Debug Build
@dotnet-bot test Outerloop NETFX x64 Debug Build

danmosemsft commented on src/System.Diagnostics.Process/tests/ProcessTests.cs:
```diff
@@ -176,13 +176,12 @@ public void TestStartWithBadWorkingDirectory()
         [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.HasWindowsShell))]
```
Do you not want this to run on Unix anymore? Because by changing HasWindowsShell to only be true on Windows, it will stop running on Unix.

danmosemsft replied:
Oh I see, you broke out a new Unix one.

maryamariyan commented:
@dotnet-bot test OSX x64 Debug Build please

maryamariyan commented:
@dotnet-bot test Outerloop Alpine.3.6 x64 Debug Build
@dotnet-bot test Outerloop CentOS.6 x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Windows x86 Debug Build
@dotnet-bot test Outerloop NETFX x64 Debug Build

maryamariyan commented:
@dotnet-bot test Outerloop Alpine.3.6 x64 Debug Build
@dotnet-bot test Outerloop CentOS.6 x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Windows x86 Debug Build
@dotnet-bot test Outerloop NETFX x64 Debug Build
